pipeline.ts

```diff
@@ -1,108 +1,73 @@
-import {
-  createRequest,
-  readReply,
-  RedisRawReply,
-  ErrorReply,
-} from "./io.ts";
-import { ErrorReplyError } from "./errors.ts";
-import { create } from "./redis.ts";
-import {
-  deferred,
-  Deferred,
-} from "./vendor/https/deno.land/std/async/mod.ts";
-import { RedisCommands } from "./command.ts";
-import { Connection, CommandExecutor } from "./connection.ts";
+import { Connection } from "./connection.ts";
+import { CommandExecutor } from "./executor.ts";
+import { RawReplyOrError, RedisRawReply, sendCommands } from "./io.ts";
+import { Redis, RedisImpl } from "./redis.ts";
+import { Deferred, deferred } from "./vendor/https/deno.land/std/async/mod.ts";
 
-const encoder = new TextEncoder();
-export type RawReplyOrError = RedisRawReply | ErrorReply;
-export type RedisPipeline = {
-  enqueue(command: string, ...args: (number | string)[]): void;
+export type RedisPipeline = Redis & {
   flush(): Promise<RawReplyOrError[]>;
-} & RedisCommands;
+};
 
 export function createRedisPipeline(
-  connection: Connection<RedisRawReply>,
-  opts?: { tx: true },
+  connection: Connection,
+  tx = false,
 ): RedisPipeline {
-  let commands: string[] = [];
-  let queue: {
-    commands: string[];
+  const executor = new PipelineExecutor(connection, tx);
+  async function flush(): Promise<RawReplyOrError[]> {
+    return executor.flush();
+  }
+  const client = new RedisImpl(connection, executor);
+  return Object.assign(client, { flush });
+}
+
+export class PipelineExecutor extends CommandExecutor {
+  private commands: {
+    command: string;
+    args: (number | string)[];
+  }[] = [];
+  private queue: {
+    commands: {
+      command: string;
+      args: (number | string)[];
+    }[];
     d: Deferred<RawReplyOrError[]>;
   }[] = [];
 
-  function dequeue() {
-    const [e] = queue;
-    if (!e) return;
-    send(e.commands)
-      .then(e.d.resolve)
-      .catch(e.d.reject)
-      .finally(() => {
-        queue.shift();
-        dequeue();
-      });
+  constructor(connection: Connection, private tx: boolean) {
+    super(connection);
   }
 
-  async function send(cmds: string[]): Promise<RawReplyOrError[]> {
-    const msg = cmds.join("");
-    await connection.writer!.write(encoder.encode(msg));
-    await connection.writer!.flush();
-    const ret: RawReplyOrError[] = [];
-    for (let i = 0; i < cmds.length; i++) {
-      try {
-        const rep = await readReply(connection.reader!);
-        ret.push(rep);
-      } catch (e) {
-        if (e instanceof ErrorReplyError) {
-          ret.push(["error", e]);
-        } else {
-          throw e;
-        }
-      }
-    }
-    return ret;
-  }
-
-  function enqueue(command: string, ...args: (number | string)[]): void {
-    const msg = createRequest(command, ...args);
-    commands.push(msg);
+  async exec(
+    command: string,
+    ...args: (string | number)[]
+  ): Promise<RedisRawReply> {
+    this.commands.push({ command, args });
```

Comment:
```suggestion
    const d = deferred<RedisRawReply>();
    this.commands.push({ command, args, d });
    return d;
```

Reply:
I copied the "OK" return from the original implementation. However as you brought this up, I'm looking at how others do this. `redis-cli` returns a status "QUEUED" as new commands are added to the pipeline. But `go-redis` returns the actual reply like you did. What do you think?

Reply:
Pipeline tests need to be updated as well

Reply:
If we return a deferred here pipeline tests need to be updated as well

Reply:
On second thought, if we return the deferred value here and user call `await` on it, the execution will be paused indefinitely because `flush()` can't be reached. So I think returning a status "QUEUED" is a good idea.

Reply:
> On second thought, if we return the deferred value here and user call await on it, the execution will be paused indefinitely because flush() can't be reached.

Oh it's not ideal.

Reply:
> So I think returning a status "QUEUED" is a good idea.

I agree with this :+1:, but I think it's out of scope of this PR.
For now, let's merge this PR!